gt.go

```diff
@@ -65,9 +65,22 @@ func (g *GT) ToBytes(e *E) []byte {
 
 // IsValid checks whether given target group element is in correct subgroup.
 func (g *GT) IsValid(e *E) bool {
-	r := g.New()
-	g.fp12.exp(r, e, qBig)
-	return r.isOne()
+	r0, r1, r2 := g.New().Set(e), g.New(), g.New()
+
+	g.fp12.frobeniusMap1(r0)
+	r1.set(r0)
+	g.fp12.frobeniusMap1(r0)
+	r2.set(r0)
+	g.fp12.frobeniusMap2(r0)
+	g.Mul(r0, r0, e)
+	if !r0.Equal(r2) {
+		return false
+	}
+	g.Exp(r0, e, bigFromHex("0xd201000000010000"))
+	g.Mul(r0, r0, r1)
+
+	return r0.IsOne()
```

Comment:
So, you are doing the cyclotomic test e·e^(p^4) == e^(p^2)  and then you test if e^p = e^u  as in Scott paper.

```suggestion
	// e^(p-u) = e^(p+1-t) == 1
	return r0.IsOne()
```